### PR TITLE
Update to arrow 32 and Switch to RawDecoder for JSON

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change

Integration test for https://github.com/apache/arrow-rs/pull/3479 and preparing for the next arrow release https://github.com/apache/arrow-rs/issues/3584

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/file_format/json.rs [hunk 1]
```diff
@@ -162,40 +154,61 @@ impl ExecutionPlan for NdJsonExec {
...
-                    let decoder = file_compression_type.convert_read(file)?;
-                    let reader = json::Reader::new(decoder, schema.clone(), options);
+                    let bytes = file_compression_type.convert_read(file)?;
+                    let reader = RawReaderBuilder::new(schema)
+                        .with_batch_size(batch_size)
+                        .build(BufReader::new(bytes))?;
                     Ok(futures::stream::iter(reader).boxed())
                 }
                 GetResult::Stream(s) => {
+                    let mut decoder = RawDecoder::try_new(schema, batch_size)?;
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/json.rs:180]: I think this interface is pretty cool, it avoids needing to scan the byte stream looking for newlines, and consequently should add some additional performance on top of the faster performance of `RawDecoder` in general

**Comment by alamb** [comment]
🥳 🦜

File: datafusion/common/src/error.rs [hunk 2]
```diff
@@ -388,46 +388,23 @@ impl DataFusionError {
...
         let mut root_error: &dyn Error = self;
-        while let Some(source) = find_source(root_error) {
+        while let Some(source) = root_error.source() {
             // walk the next level
             root_error = source;
             // remember the lowest datafusion error so far
             if let Some(e) = root_error.downcast_ref::<DataFusionError>() {
                 last_datafusion_error = e;
+            } else if let Some(e) = root_error.downcast_ref::<Arc<DataFusionError>>() {
+                // As `Arc<T>::source()` calls through to `T::source()` we need to
```
Comment (alamb) [inline datafusion/common/src/error.rs:398]: 👍

File: datafusion/core/src/avro_to_arrow/schema.rs [hunk 3]
```diff
@@ -217,6 +217,9 @@ fn default_field_name(dt: &DataType) -> &str {
         DataType::Union(_, _, _) => "union",
         DataType::Dictionary(_, _) => "map",
         DataType::Map(_, _) => unimplemented!("Map support not implemented"),
+        DataType::RunEndEncoded(_, _) => {
+            unimplemented!("RunEndEncoded support not implemented")
```
Comment (alamb) [inline datafusion/core/src/avro_to_arrow/schema.rs:221]: image

File: datafusion/core/src/physical_plan/file_format/csv.rs [hunk 4]
```diff
@@ -21,7 +21,6 @@ use crate::datasource::file_format::file_type::FileCompressionType;
 use crate::error::{DataFusionError, Result};
 use crate::execution::context::{SessionState, TaskContext};
 use crate::physical_plan::expressions::PhysicalSortExpr;
-use crate::physical_plan::file_format::delimited_stream::newline_delimited_stream;
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/csv.rs:24]: Can the corresponding `newline_delimited_stream` module be deleted too? 

https://github.com/search?q=repo%3Aapache%2Farrow-datafusion%20newline_delimited_stream&type=code

File: datafusion/core/src/physical_plan/file_format/csv.rs [hunk 5]
```diff
@@ -224,20 +240,38 @@ impl FileOpener for CsvOpener {
...
+                                    Some(Ok(b)) => b,
+                                    Some(Err(e)) => {
+                                        return Poll::Ready(Some(Err(e.into())))
+                                    }
+                                    None => break,
+                                };
+                            }
+                            let decoded = match decoder.decode(buffered.as_ref()) {
+                                // Note: the decoder needs to be called with an empty
+                                // array to delimt the final record
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/csv.rs:264]:
```suggestion
                                // array to delimit the final record
```

Comment (alamb) [reply datafusion/core/src/physical_plan/file_format/csv.rs:264]: I must be missing how the code is called with an empty buffer. If all data in buffered was consumed and then the next poll was empty, won't that break out of the the loop prior to calling `decode()` 🤔

File: datafusion/proto/src/logical_plan/to_proto.rs [hunk 6]
```diff
@@ -218,7 +218,7 @@ impl TryFrom<&DataType> for protobuf::arrow_type::ArrowTypeEnum {
             DataType::Decimal256(_, _) => {
                 return Err(Error::General("Proto serialization error: The Decimal256 data type is not yet supported".to_owned()))
             }
-            DataType::Map(_, _) => {
+            DataType::Map(_, _) | DataType::RunEndEncoded(_, _) => {
```
Comment (alamb) [inline datafusion/proto/src/logical_plan/to_proto.rs:221]: I recommend either updating the error message here or adding a separate clause for RunEndEncoded

**Review by alamb — APPROVED** [review]
Looks great @tustvold  -- the only question I have is on feeding in empty buffers to the csv reader -- but perhaps I am misreading something

File: datafusion/core/src/physical_plan/file_format/csv.rs (hunk 4 above)
Comment (tustvold) [reply datafusion/core/src/physical_plan/file_format/csv.rs:24]: Unfortunately it is still used by the schema inference logic, I'll see about resurrecting the PR to move to using the upstream implementation

File: datafusion/core/src/physical_plan/file_format/csv.rs (hunk 5 above)
Comment (tustvold) [reply datafusion/core/src/physical_plan/file_format/csv.rs:264]: You are quite correct, I'm investigating how this is working...

**Pull request merged**